### PR TITLE
Update api beta test guide: add 'editoffer'

### PR DESCRIPTION
**Pull request by ghubstan**
This is the 5th in a series of PRs, starting at https://github.com/bisq-network/bisq/pull/5570
https://github.com/bisq-network/bisq/pull/5576 Should be reviewed/merged before this one.

### CODE REVIEW

File: apitest/src/test/java/bisq/apitest/method/offer/CreateOfferUsingMarketPriceMarginTest.java [hunk 1]
```diff
@@ -220,6 +233,41 @@ public void testCreateBRLBTCSellOffer6Point55PctPriceMargin() {
...
+                MAKER_FEE_CURRENCY_CODE,
+                triggerPriceAsLong);
+        genBtcBlocksThenWait(1, 4000); // give time to add to offer book
+        newOffer = aliceClient.getMyOffer(newOffer.getId());
+        log.info("OFFER #5:\n{}", formatOfferTable(singletonList(newOffer), "usd"));
+        assertEquals(triggerPriceAsLong, newOffer.getTriggerPrice());
+    }
+
+    public static void main(String[] args) {
+        // TODO DELETE ME
```
Comment (sqrrm) [inline apitest/src/test/java/bisq/apitest/method/offer/CreateOfferUsingMarketPriceMarginTest.java]: Looks like this should be deleted if I'm to trust the comments...

**Review by sqrrm — COMMENTED** [review]
Some TODOs to look at, but overall the PR looks good.

**Review by sqrrm — COMMENTED** [review]
There were some other TODOs I didn't comment on. Is this the intended final version of the PR?

**Comment by ghubstan** [comment]
> Is this the intended final version of the PR?

Yes.  It is the final PR of the series:
https://github.com/bisq-network/bisq/pull/5571
https://github.com/bisq-network/bisq/pull/5572
https://github.com/bisq-network/bisq/pull/5576
https://github.com/bisq-network/bisq/pull/5577

I am going through all the commits to find/fix those TODOs.

**Comment by ripcurlx** [comment]
@ghubstan I'm walking through the documentation right now trying out all scenarios and just recognized that enable/disable doesn't seem to work
```
➜  bisq git:(05-update-api-beta-test-guide) ./bisq-cli --password=mySecret editoffer --offer-id=746908-1d802650-45c0-4007-afac-f9a8a6d6d0af-165 --enabled=false
Error: enabled is not a recognized option

```

**Comment by ripcurlx** [comment]
> @ghubstan I'm walking through the documentation right now trying out all scenarios and just recognized that enable/disable doesn't seem to work
> 
> ```
> ➜  bisq git:(05-update-api-beta-test-guide) ./bisq-cli --password=mySecret editoffer --offer-id=746908-1d802650-45c0-4007-afac-f9a8a6d6d0af-165 --enabled=false
> Error: enabled is not a recognized option
> ```

There is a type in the docs `enabled` should be `enable`. Using it like this enabling/disabling works.

File: apitest/docs/api-beta-test-guide.md [hunk 2]
```diff
@@ -408,8 +408,118 @@ The offer will be removed from other Bisq users' offer views, and paid transacti
...
+
+#### Enable and Disable Offer
+
+Existing offers you create can be disabled (removed from offer book) and re-enabled (re-published to offer book).
+
+To disable an offer:
+```
+./bisq-cli --password=xyz --port=9998 editoffer \
+    --offer-id=83e8b2e2-51b6-4f39-a748-3ebd29c22aea \
+    --enabled=false
```
Comment (ripcurlx) [inline apitest/docs/api-beta-test-guide.md]: `--enabled=false` should be `--enable=false`

File: apitest/docs/api-beta-test-guide.md [hunk 3]
```diff
@@ -408,8 +408,118 @@ The offer will be removed from other Bisq users' offer views, and paid transacti
...
+```
+
+To enable an offer:
+```
+./bisq-cli --password=xyz --port=9998 editoffer \
+    --offer-id=83e8b2e2-51b6-4f39-a748-3ebd29c22aea \
+    --enabled=true
```
Comment (ripcurlx) [inline apitest/docs/api-beta-test-guide.md]: `--enabled=true` should be `--enable=true`

File: apitest/docs/api-beta-test-guide.md (hunk 2 above)
Comment (ghubstan) [reply apitest/docs/api-beta-test-guide.md]: Fixed in commit a4278a4147c78ae7e4350058457af59a0083c9de.

File: apitest/docs/api-beta-test-guide.md (hunk 3 above)
Comment (ghubstan) [reply apitest/docs/api-beta-test-guide.md]: Fixed in commit a4278a4147c78ae7e4350058457af59a0083c9de.

**Comment by ripcurlx** [comment]
@ghubstan From a usage perspective everything works as expected - great work! The only thing I recognized sometimes (unfortunately not reproducable yet) is, that after the editing the offer is not shown in the offerbook of the other user. If you switch between tabs it is shown again. I'm not sure if there is some weird UI race condition in some cases. Did you experience something similar in your tests?

**Comment by ripcurlx** [comment]
When trying to use the API programatic I recognized that it is not 100% straight forward working with the API response. E.g. parsing through lines like this `'Sell BSQ (Buy BTC)              0.00003800      0,02000000           526.32  Altcoins        2021-07-13T09:31:56Z  IRDQJ-30666af2-460b-4f23-94a5-ec632913b8d0-165      ` You can't simply split for spaces as the first value contains spaces as well. Maybe we could improve the layout so it is easier to work with (provide JSON response by using a parameter?)

**Comment by ripcurlx** [comment]
I just recognized when creating a new offer in the response the enabled state is displayed wrong:
```
 ./bisq-cli --password=mySecret createoffer --payment-account=f9a5ce72-6496-4244-a0f8-a49b33bf121b --direction=BUY --currency-code=BSQ --amount=0.01 --fixed-price=0.00004 --security-deposit=15.0 --fee-currency=BSQ
Enabled  Buy/Sell            Price in BTC for 1 BSQ  BTC(min - max)   BSQ(min - max)  Payment Method    Creation Date (UTC)   ID
NO       Sell BSQ (Buy BTC)              0.00004000      0,01000000           250.00  Altcoins Instant  2021-07-15T07:49:52Z  CgtPBay9-7826b4d0-7da6-419c-a70b-657934016485-165   
```
It says NO, but it is visible to the other peer already. When requesting the state of the specific offer it does display the correct value. 
```
 bisq git:(05-update-api-beta-test-guide) ./bisq-cli --password=mySecret getmyoffer --offer-id=CgtPBay9-7826b4d0-7da6-419c-a70b-657934016485-165
Enabled  Buy/Sell            Price in BTC for 1 BSQ  BTC(min - max)   BSQ(min - max)  Payment Method    Creation Date (UTC)   ID
YES      Sell BSQ (Buy BTC)              0.00004000      0,01000000           250.00  Altcoins Instant  2021-07-15T07:49:52Z  CgtPBay9-7826b4d0-7da6-419c-a70b-657934016485-165```

**Comment by ripcurlx** [comment]
As I just recognized it:
```
Enabled  Buy/Sell            Price in BTC for 1 BSQ  BTC(min - max)   BSQ(min - max)  Payment Method    Creation Date (UTC)   ID
YES      Sell BSQ (Buy BTC)              0.00003700      0,01000000           270.27  Altcoins Instant  2021-07-15T07:48:37Z  1GEEMUWJ-eff9e2a0-0550-4c44-aec7-6ead4b9d387b-165   
YES      Sell BSQ (Buy BTC)              0.00003700      0,01000000           270.27  Altcoins Instant  2021-07-15T07:49:52Z  CgtPBay9-7826b4d0-7da6-419c-a70b-657934016485-165   
YES      Sell BSQ (Buy BTC)              0.00003700      0,10000000         2,702.70  Altcoins          2021-07-13T08:22:33Z  746908-1d802650-45c0-4007-afac-f9a8a6d6d0af-165
```
Is there a reason why the decimal point for the price is `.` and for amount `,` ?

**Comment by ghubstan** [comment]
> @ghubstan From a usage perspective everything works as expected - great work! The only thing I recognized sometimes (unfortunately not reproducable yet) is, that after the editing the offer is not shown in the offerbook of the other user. If you switch between tabs it is shown again. I'm not sure if there is some weird UI race condition in some cases. Did you experience something similar in your tests?

>>  Did you experience something similar in your tests?  
I did not, only  random delays for the republishing of modified offers.

An edited offer gets republished: it is removed from the offer book, then re-added with your changes. The delay between removal and re-publishing can be some seconds.   The next time it disappears from the other user's offerbook, wait for it to appear.  You shouldn't have to navigate away then back.  (But if you disabled it, it would not return to the other user's offer book until after you re-enable it.  I doubt that's the case, but it's possible to forget using `enable=false` while doing something else too.)

[CORRECTION]

I am seeing Alice's edited offer disappearing from Bob's desktop, forcing me to navigate away and return to make it reappear.  Not sure what the solution is yet.

**Comment by ghubstan** [comment]
> I just recognized when creating a new offer in the response the enabled state is displayed wrong:

Some async tasks (fully preparing and publishing a new offer) do not finish in time to be included in the server's response to the CLI.  The `OpenOffer.State {AVAILABLE, ... DEACTIVATED)` is one bit of information the CLI does not get from the server in the `createoffer` response because the new `OpenOffer` instance has not yet been added to the `OpenOfferManager`.   

I didn't want to lie about the true `Enabled` state in the `createoffer` response, that's why you see `NO`.
But as you see,  calling `getmyoffer` right after `createoffer` shows the correct `Enabled` state `YES`.



> It says NO, but it is visible to the other peer already. When requesting the state of the specific offer it does display the correct value.

The CLI is session-less, it only knows that at the instance of offer creation, no `OpenOffer.STATE` is ready to be passed to the CLI.  By the time you become annoyed at seeing `Enabled`=`NO`, the `OpenOffer` has been added to the `OpenOfferManager` and is published to other peers.  About 1-3 seconds after you successfully call `createoffer`, the correct `Enabled` value shows up in your `getmyoffer` console output.

Not great.  Would you prefer that the CLI block on `createoffer` until the `OpenOfferManager` has the new `OpenOffer` object?  That did not seem like a good idea before now, and still doesn't... what do you think?

**Comment by ghubstan** [comment]
> When trying to use the API programatic I recognized that it is not 100% straight forward working with the API response. E.g. parsing through lines like this `'Sell BSQ (Buy BTC) 0.00003800 0,02000000 526.32 Altcoins 2021-07-13T09:31:56Z IRDQJ-30666af2-460b-4f23-94a5-ec632913b8d0-165 ` You can't simply split for spaces as the first value contains spaces as well. Maybe we could improve the layout so it is easier to work with (provide JSON response by using a parameter?)

To justify and align column values, I have to add quite a bit of right-padding and left-padding, and that's what's making parsing so difficult.

@dmos62 and I discussed this... he also wanted a more machine friendly CLI console output.  I think Json is the answer.  But that's another project/PR in itself and I don't think it should be rushed into this PR.

I work around this problem in the apitest trading simulation scripts:  Look at line 234 in `trade-simulation-utils.sh`.

```
gettradedetail() {
    TRADE_DESC="$1"
    # Get 2nd line of gettrade cmd output, and squeeze multi space delimiters into one space.
    TRADE_DETAIL=$(echo "$TRADE_DESC" | sed -n '2p' | tr -s ' ')
    commandalert $? "Could not get trade detail (line 2 of gettrade output)."
    echo "$TRADE_DETAIL"
}
```

The bash function `gettradedetail` takes one parameter, the full, raw `gettrade` CLI output  (the header and values rows).
That raw `TRADE_DESC` variable gets squeezed into `TRADE_DETAIL`, where multi-space delimiters are replaced by single space delimiters.

**Comment by ghubstan** [comment]
> Is there a reason why the decimal point for the price is `.` and for amount `,` ?

That's a bug.  Looking into it.

649c98a should fix this for you

**Comment by ripcurlx** [comment]
> > I just recognized when creating a new offer in the response the enabled state is displayed wrong:
> 
> Some async tasks (fully preparing and publishing a new offer) do not finish in time to be included in the server's response to the CLI. The `OpenOffer.State {AVAILABLE, ... DEACTIVATED)` is one bit of information the CLI does not get from the server in the `createoffer` response because the new `OpenOffer` instance has not yet been added to the `OpenOfferManager`.
> 
> I didn't want to lie about the true `Enabled` state in the `createoffer` response, that's why you see `NO`.
> But as you see, calling `getmyoffer` right after `createoffer` shows the correct `Enabled` state `YES`.
> 
> > It says NO, but it is visible to the other peer already. When requesting the state of the specific offer it does display the correct value.
> 
> The CLI is session-less, it only knows that at the instance of offer creation, no `OpenOffer.STATE` is ready to be passed to the CLI. By the time you become annoyed at seeing `Enabled`=`NO`, the `OpenOffer` has been added to the `OpenOfferManager` and is published to other peers. About 1-3 seconds after you successfully call `createoffer`, the correct `Enabled` value shows up in your `getmyoffer` console output.
> 
> Not great. Would you prefer that the CLI block on `createoffer` until the `OpenOfferManager` has the new `OpenOffer` object? That did not seem like a good idea before now, and still doesn't... what do you think?

Ok - I do understand the reasoning behind this, but from a developer perspective looking at the response I'd think I have to do an additional enable after creating the offer to get it publish. Maybe we could add another state like `PENDING`? WDYT?

**Comment by ripcurlx** [comment]
> > When trying to use the API programatic I recognized that it is not 100% straight forward working with the API response. E.g. parsing through lines like this `'Sell BSQ (Buy BTC) 0.00003800 0,02000000 526.32 Altcoins 2021-07-13T09:31:56Z IRDQJ-30666af2-460b-4f23-94a5-ec632913b8d0-165 ` You can't simply split for spaces as the first value contains spaces as well. Maybe we could improve the layout so it is easier to work with (provide JSON response by using a parameter?)
> 
> To justify and align column values, I have to add quite a bit of right-padding and left-padding, and that's what's making parsing so difficult.
> 
> @dmos62 and I discussed this... he also wanted a more machine friendly CLI console output. I think Json is the answer. But that's another project/PR in itself and I don't think it should be rushed into this PR.
> 
> I work around this problem in the apitest trading simulation scripts: Look at line 234 in `trade-simulation-utils.sh`.
> 
> ```
> gettradedetail() {
>     TRADE_DESC="$1"
>     # Get 2nd line of gettrade cmd output, and squeeze multi space delimiters into one space.
>     TRADE_DETAIL=$(echo "$TRADE_DESC" | sed -n '2p' | tr -s ' ')
>     commandalert $? "Could not get trade detail (line 2 of gettrade output)."
>     echo "$TRADE_DETAIL"
> }
> ```
> 
> The bash function `gettradedetail` takes one parameter, the full, raw `gettrade` CLI output (the header and values rows).
> That raw `TRADE_DESC` variable gets squeezed into `TRADE_DETAIL`, where multi-space delimiters are replaced by single space delimiters.

Yes, I was able to work with the current format as well, it is just an additional mental hurdle for devs playing around with the API. Sure, no need to squeeze JSON support into this PR as well, but I think it would make the API way more accessible for other devs trying to implement their use cases.

**Review by sqrrm — DISMISSED** [review]
utACK

@ripcurlx Do you have any more comments on this PR?

**Comment by sqrrm** [comment]
I don't think the api calls should block. Better if the output would make it clearer that request is being processed, maybe by having a `target state` or something like that to indicate that something is happening.

**Comment by ghubstan** [comment]
> Ok - I do understand the reasoning behind this, but from a developer perspective looking at the response I'd think I have to do an additional enable after creating the offer to get it publish. Maybe we could add another state like `PENDING`? WDYT?

`PENDING` in this particular scenario is good.   Thanks for such a simple, self-explanatory solution.

**Comment by dmos62** [comment]
> 
> 
> > Ok - I do understand the reasoning behind this, but from a developer perspective looking at the response I'd think I have to do an additional enable after creating the offer to get it publish. Maybe we could add another state like `PENDING`? WDYT?
> 
> `PENDING` in this particular scenario is good. Thanks for such a simple, self-explanatory solution.

Sorry to chime in. What does the Enabled column signify? It sounds like it might functionally mean something different than it does semantically. You wouldn't expect it to go from disabled to enabled by itself. Maybe it's a synonym for published? Could there be a semantic conflict in there?

PS: bravo for the work you've done @ghubstan

**Comment by ghubstan** [comment]
>>  What does the Enabled column signify? 

Hey D, thanks for the nice comment.

From the UI's open offers view, you can enable/disable an offer -- add/remove the offer from the offer book.
The api's `editoffer` command has an `--enable=true|false` option to do the same thing.

**Comment by dmos62** [comment]
> 
> 
> > > What does the Enabled column signify?
> 
> Hey D, thanks for the nice comment.
> 
> From the UI's open offers view, you can enable/disable an offer -- add/remove the offer from the offer book.
> The api's `editoffer` command has an `--enable=true|false` option to do the same thing.

Sounds like there's two state variables (at least implicitly). One is a settable variable signifying the state we request the offer to be in, and the other is a read-only variable for what state it actually is in. Thought it might be worth pointing out.

The `PENDING` solution sounds good. The column is effectively the read-only variable.

**Comment by ghubstan** [comment]
test

**Comment by ghubstan** [comment]
File conflicts resolved in child PR https://github.com/bisq-network/bisq/pull/5666

**Pull request merged**